Semantic Segmentation Train Evaluation

Streakfull opened the pull request:
## In this PR,
The main idea is to log the semantic segmentation (IOU, Accuracy) metrics during training so that they can be compared with the validation metrics in addition to the loss.

### Changes
* Added a `SemSegEvaluatorTrain` hook that writes the training metrics to the tensorboard and wandb.
* Added the hook in the default configuration for `scannetpp` methods.
* Semantic segmenation models now also return the `seg_logits` during training for evaluation.
---
### Demo
![image](https://github.com/user-attachments/assets/21468552-bb25-4647-af0d-797f19df8ad4)

The full logs can be found [here](https://api.wandb.ai/links/streakfull-technical-university-of-munich/ed0erarb).
---
*Note this PR depends on another open PR, which can be found [here](https://github.com/Pointcept/Pointcept/pull/413)*.

Gofinge commented:
> ## In this PR,
> The main idea is to log the semantic segmentation (IOU, Accuracy) metrics during training so that they can be compared with the validation metrics in addition to the loss.
> 
> ### Changes
> * Added a `SemSegEvaluatorTrain` hook that writes the training metrics to the tensorboard and wandb.
> * Added the hook in the default configuration for `scannetpp` methods.
> * Semantic segmenation models now also return the `seg_logits` during training for evaluation.
> 
> ### Demo
> ![image](https://private-user-images.githubusercontent.com/34299200/426835759-21468552-bb25-4647-af0d-797f19df8ad4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI5NjA5MzEsIm5iZiI6MTc0Mjk2MDYzMSwicGF0aCI6Ii8zNDI5OTIwMC80MjY4MzU3NTktMjE0Njg1NTItYmIyNS00NjQ3LWFmMGQtNzk3ZjE5ZGY4YWQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI2VDAzNDM1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2NzE3NTdmZjhkZjc3MTk1M2FmOWRiMDBlNDVlMTE2ZTIwMTY3NDgzODEzY2I5M2E3YjE5ZDY4OTU0MzAzZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.i5i12YouT846YmcthnklUZOBazfYfY4iJc_jGcluYn8)
> 
> The full logs can be found [here](https://api.wandb.ai/links/streakfull-technical-university-of-munich/ed0erarb).
> 
> _Note this PR depends on another open PR, which can be found [here](https://github.com/Pointcept/Pointcept/pull/413)_.

Looks awesome

Gofinge commented:
@Streakfull 
From now on, one day on pr. I will process this tmr.

Streakfull commented:
@Gofinge Alright, sounds good thanks! We ran into an issue today, when trying to reproduce some evaluation results with stored checkpoints for `configs/scannetpp/semseg-spunet-v1m1-0-base.py` , not sure yet, what else isn't working. This is the issue:
* Running the evaluation on commit (`e4de3c25f57d0625cdcb66589cf180e838a05b19`)[https://github.com/Pointcept/Pointcept/commit/e4de3c25f57d0625cdcb66589cf180e838a05b19] give 0.478 mIou (this is correct and matches our previous results in the fork)
* Running the evaluation in the latest branch gives 0.4 mIou
* After some debugging, I was able to narrow it down to differences in the `datasets/transform` file. I.E I replaced the `datasets/transform` file in the working branch with the latest `datasets/transform` file, with the latest file and adjusted the configuration accordingly, then got the 0.4 mIou.

Do you have any suggestions on how to proceed with this? Should I open a separate issue?

RayYoh commented:
> @Gofinge Alright, sounds good thanks! We ran into an issue today, when trying to reproduce some evaluation results with stored checkpoints for `configs/scannetpp/semseg-spunet-v1m1-0-base.py` , not sure yet, what else isn't working. This is the issue:
> 
> * Running the evaluation on commit [`e4de3c25f57d0625cdcb66589cf180e838a05b19`](https://github.com/Pointcept/Pointcept/commit/e4de3c25f57d0625cdcb66589cf180e838a05b19) gives 0.478 mIou (this is correct and matches our previous results in the fork)
> * Running the evaluation in the latest main branch gives 0.4 mIou
> * After some debugging, I was able to narrow it down to differences in the `datasets/transform` file. I.E I replaced the `datasets/transform` file in the working branch with the latest `datasets/transform` file  and adjusted the configuration accordingly, then got the 0.4 mIou.
> 
> Do you have any suggestions on how to proceed with this? Should I open a separate issue?

The change I think maybe `NormalizeColor`. Previous range is [-1, 1] but the latest version is [0,1].

This snapshot is from previous version.
![image](https://github.com/user-attachments/assets/2b826850-09e5-4301-8051-2ea06e1e9e85)

Gofinge commented:
Thanks @RayYoh
